File: gdx/test/com/badlogic/gdx/utils/JsonValueTest.java
package com.badlogic.gdx.utils;

import org.junit.Assert;
import org.junit.Test;

public class JsonValueTest {

    @Test
    public void testAddingRemovedValue() {
        // Prepare two JSON objects
        JsonValue firstObject = new JsonValue(JsonValue.ValueType.object);
        JsonValue secondObject = new JsonValue(JsonValue.ValueType.object);

        firstObject.addChild("a", new JsonValue("A"));
        secondObject.addChild("b", new JsonValue("B"));
        secondObject.addChild("c", new JsonValue("C"));

        // Remove an item from one object and add it to the other
        JsonValue b = secondObject.remove("b");
        firstObject.addChild(b);

        // Check if both objects have the expected children
        Assert.assertNotNull(firstObject.get("a"));
        Assert.assertNotNull(firstObject.get("b"));
        Assert.assertNull(firstObject.get("c"));

        Assert.assertNull(secondObject.get("a"));
        Assert.assertNull(secondObject.get("b"));
        Assert.assertNotNull(secondObject.get("c"));
    }

    @Test
    public void testReplaceValue() {
        JsonValue object = new JsonValue(JsonValue.ValueType.object);

        object.addChild("a", new JsonValue("A"));
        object.addChild("b", new JsonValue("B"));
        object.addChild("c", new JsonValue("C"));
        object.addChild("d", new JsonValue("D"));

        object.setChild("b", new JsonValue("X"));
        object.setChild("c", new JsonValue("Y"));

        Assert.assertEquals("A", object.get("a").asString());
        Assert.assertEquals("X", object.get("b").asString());
        Assert.assertEquals("Y", object.get("c").asString());
        Assert.assertEquals("D", object.get("d").asString());

        object.setChild("a", new JsonValue("W"));
        JsonValue z = new JsonValue("Z");
        z.name = "d";
        object.setChild(z);

        Assert.assertEquals(object.get("a").parent(), object);
        Assert.assertEquals("W", object.get("a").asString());
        Assert.assertEquals("X", object.get("b").asString());
        Assert.assertEquals("Y", object.get("c").asString());
        Assert.assertEquals("Z", object.get("d").asString());
    }

    @Test
    public void testCopyConstructor() {
        JsonValue b = new JsonValue(JsonValue.ValueType.object);
        b.addChild("c", new JsonValue("C"));
        b.addChild("d", new JsonValue("D"));

        JsonValue object = new JsonValue(JsonValue.ValueType.object);
        object.addChild("a", new JsonValue("A"));
        object.addChild("b", b);
        object.addChild("e", new JsonValue(123));

        JsonValue copy = new JsonValue(object);

        Assert.assertEquals("A", object.get("a").asString());
        Assert.assertEquals(object.get("b"), b);
        Assert.assertEquals("C", object.get("b").get("c").asString());
        Assert.assertEquals("D", object.get("b").get("d").asString());
        Assert.assertEquals(object.get("b").parent(), object);
        Assert.assertEquals(123, object.get("e").asInt());
        Assert.assertNull(object.get("E"));
        Assert.assertEquals(123, object.getIgnoreCase("E").asInt());

        Assert.assertEquals("A", copy.get("a").asString());
        Assert.assertNotEquals(copy.get("b"), b);
        Assert.assertEquals("C", copy.get("b").get("c").asString());
        Assert.assertEquals("D", copy.get("b").get("d").asString());
        Assert.assertEquals(copy.get("b").parent(), copy);
        Assert.assertEquals(123, copy.get("e").asInt());
        Assert.assertNull(copy.get("E"));
        Assert.assertEquals(123, copy.getIgnoreCase("E").asInt());

        JsonValue bCopy = new JsonValue(copy.get("b"));
        Assert.assertNotEquals(b, bCopy);
        Assert.assertNotEquals(copy.get("b"), bCopy);
        Assert.assertEquals("C", bCopy.get("c").asString());
        Assert.assertEquals("D", bCopy.get("d").asString());
        Assert.assertEquals(bCopy.get("d").parent(), bCopy);
    }
}

File: gdx/test/com/badlogic/gdx/utils/JsonValueTest.kt
package com.badlogic.gdx.utils

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test

class JsonValueTest {
    @Test
    fun testAddingRemovedValue() {
        // Prepare two JSON objects
        val firstObject = JsonValue(JsonValue.ValueType.objectValue)
        val secondObject = JsonValue(JsonValue.ValueType.objectValue)

        firstObject.addChild("a", JsonValue("A"))
        secondObject.addChild("b", JsonValue("B"))
        secondObject.addChild("c", JsonValue("C"))

        // Remove an item from one object and add it to the other
        val b = secondObject.remove("b")
        firstObject.addChild(b)

        // Check if both objects have the expected children
        assertNotNull(firstObject.get("a"))
        assertNotNull(firstObject.get("b"))
        assertNull(firstObject.get("c"))

        assertNull(secondObject.get("a"))
        assertNull(secondObject.get("b"))
        assertNotNull(secondObject.get("c"))
    }

    @Test
    fun testReplaceValue() {
        val jsonObject = JsonValue(JsonValue.ValueType.objectValue)

        jsonObject.addChild("a", JsonValue("A"))
        jsonObject.addChild("b", JsonValue("B"))
        jsonObject.addChild("c", JsonValue("C"))
        jsonObject.addChild("d", JsonValue("D"))

        jsonObject.setChild("b", JsonValue("X"))
        jsonObject.setChild("c", JsonValue("Y"))

        assertEquals("A", jsonObject.get("a").asString())
        assertEquals("X", jsonObject.get("b").asString())
        assertEquals("Y", jsonObject.get("c").asString())
        assertEquals("D", jsonObject.get("d").asString())

        jsonObject.setChild("a", JsonValue("W"))
        val z = JsonValue("Z")
        z.name = "d"
        jsonObject.setChild(z)

        assertEquals(jsonObject.get("a").parent(), jsonObject)
        assertEquals("W", jsonObject.get("a").asString())
        assertEquals("X", jsonObject.get("b").asString())
        assertEquals("Y", jsonObject.get("c").asString())
        assertEquals("Z", jsonObject.get("d").asString())
    }

    @Test
    fun testCopyConstructor() {
        val b = JsonValue(JsonValue.ValueType.objectValue)
        b.addChild("c", JsonValue("C"))
        b.addChild("d", JsonValue("D"))

        val `object` = JsonValue(JsonValue.ValueType.objectValue)
        `object`.addChild("a", JsonValue("A"))
        `object`.addChild("b", b)
        `object`.addChild("e", JsonValue(123))

        val copy = JsonValue(`object`)

        assertEquals("A", `object`.get("a").asString())
        assertEquals(`object`.get("b"), b)
        assertEquals("C", `object`.get("b").get("c").asString())
        assertEquals("D", `object`.get("b").get("d").asString())
        assertEquals(`object`.get("b").parent(), `object`)
        assertEquals(123, `object`.get("e").asInt().toLong())
        assertNull(`object`.get("E"))
        assertEquals(123, `object`.getIgnoreCase("E").asInt().toLong())

        assertEquals("A", copy.get("a").asString())
        assertNotEquals(copy.get("b"), b)
        assertEquals("C", copy.get("b").get("c").asString())
        assertEquals("D", copy.get("b").get("d").asString())
        assertEquals(copy.get("b").parent(), copy)
        assertEquals(123, copy.get("e").asInt().toLong())
        assertNull(copy.get("E"))
        assertEquals(123, copy.getIgnoreCase("E").asInt().toLong())

        val bCopy = JsonValue(copy.get("b"))
        assertNotEquals(b, bCopy)
        assertNotEquals(copy.get("b"), bCopy)
        assertEquals("C", bCopy.get("c").asString())
        assertEquals("D", bCopy.get("d").asString())
        assertEquals(bCopy.get("d").parent(), bCopy)
    }
}
